### PR TITLE
Core: create the per world random object in the world constructor

### PR DESCRIPTION
**Pull request by alwaysintreble**
## What is this fixing or adding?
Honestly was stupid to not just do this in the first place when implementing random as an attr on the world itself. This does make seeds generated before this change different, but plan was to eventually drop `per_slot_randoms` completely anyways, so this just moves forward that particular breakage.
Main reason for doing this is because item link worlds don't have a random object associated so trying to use their random object during filler creation causes a crash. That in particular only affects the messenger and lufia2 but is easy to miss so could affect more on the future. @el-u so you're aware that Lufia currently crashes with item links without this change. 

## How was this tested?
Unit tests. More practical testing couldn't hurt, but this is relatively low impact.

### CODE REVIEW

**Comment by alwaysintreble** [comment]
This is now reliant on #2290

**Review by BadMagic100 — APPROVED** [review]
Looks good to me. I also wouldn't mind keeping the random assignment in set_seed and just assigning directly to world randoms instead of going via the dict but I don't really have a strong preference

**Comment by black-sliver** [comment]
> I also wouldn't mind keeping the random assignment in set_seed

I think having random available asap is a good thing, otherwise people may try to hack their way around it. The only downside i see is that you can not create a world before `set_seed()` now. I'm wondering If this should be detected - i.e. throw an exception in `set_seed()` if worlds are already constructed.

> alwaysintreble added waiting-on: core-review

Tests are failing.

**Comment by alwaysintreble** [comment]
>  I'm wondering If this should be detected - i.e. throw an exception in `set_seed()` if worlds are already constructed.

Done.

**Comment by BadMagic100** [comment]
Regarding testing (once 2290 is merged) I would actually expect all the relevant code paths to be hit by automated tests here; since the change relates to multiworld instantiation even untested/poorly tested worlds should be ok because they won't be instantiating multiworld instances, or other worlds

File: test/general/__init__.py [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
 from BaseClasses import MultiWorld, CollectionState
 from worlds.AutoWorld import call_all, World
 
 gen_steps = ("generate_early", "create_regions", "create_items", "set_rules", "generate_basic", "pre_fill")
 
 
-def setup_solo_multiworld(world_type: Type[World], steps: Tuple[str, ...] = gen_steps) -> MultiWorld:
+def setup_solo_multiworld(
+    world_type: Type[World], steps: Tuple[str, ...] = gen_steps, seed: Optional[int] = None
+) -> MultiWorld:
```
Comment (black-sliver) [inline test/general/__init__.py:12]: Did you forget to use seed in set_seed() below?

Comment (alwaysintreble) [reply test/general/__init__.py:12]: ~~no i just really hate the way dlcq and sdv do their tests~~

**Review by black-sliver — APPROVED** [review]
I'm fairly certain this is gonna break tests of worlds that have not been merged (or PR'd) yet, but I don't see a way to make this more back compatible, without adding duplicate world.random init, so we should just merge this.

**Pull request merged**